### ISSUE TITLE
Reach and Hitboxes Bypass through exploitation of the Movement Threshold

### ISSUE THREAD

**Issue by Hwaanx** [GrimAnticheat/Grim#1212]
### Describe the bypass and how to replicate it

https://www.youtube.com/watch?v=KFp5DhlXh4Y

Afaik the hitboxes of the player are expanded on all sides because of the movement threshold. 
It should also take account the players movement direction. The hitbox shouldn't expand to -X and +Z when the player is moving to +X and -Z coordinates.

Client Version: 1.16.5 but joined with 1.17.1 with Via

### Grim version

2.3.46

### Server version

1.19.4 Paper

### Plugins

FAWE, GrimAC, SkinsRestorer, ViaBackwards, ViaVersion

**Comment by mc-zhonghuang** [comment]
In fact, hitboxes should be the smooth value between the position updated by the previous tick and the position updated by this tick.
But I have a question, if we want to perform such calculations on the server, won't the computational load be too large, resulting in a decrease in the TPS value of some servers that are not very good.

**Comment by Hwaanx** [comment]
> In fact, hitboxes should be the smooth value between the position updated by the previous tick and the position updated by this tick. But I have a question, if we want to perform such calculations on the server, won't the computational load be too large, resulting in a decrease in the TPS value of some servers that are not very good.

I don't know what will be the cost of calculating this but it causes significant bypasses, so it should get added and if its too heavy on the server it can be an optional setting

**Comment by MWHunter** [comment]
This is from the idle packet, it's very difficult to reach check 1.9+ players when they aren't moving since mojang removed the ability for servers to know when the client ticked if the player isn't moving...

**Comment by Hwaanx** [comment]
> This is from the idle packet, it's very difficult to reach check 1.9+ players when they aren't moving since mojang removed the ability for servers to know when the client ticked if the player isn't moving...

This needs a fix though isn't there an easy way?

**Comment by ghost** [comment]
> > This is from the idle packet, it's very difficult to reach check 1.9+ players when they aren't moving since mojang removed the ability for servers to know when the client ticked if the player isn't moving...
> 
> This needs a fix though isn't there an easy way?

The way grim checking for reach is all ok but there is many flaws. For example grim save player location by using playerAttackQueue. And it going to run tickFlying() to check and flag. But the problem is grim clear entire list after running the check. And tickFlying() only run when it receiving flying packets. So to bypass you just need to get the first hit in the range of 3 blocks and then cancel movement packets right after. Grim will not run the checks on you and let you be allowing you to 6 blocks reach (you can't move tho :) ). In 1.8 you need to code something to manually do this for you prob? But in 1.9 cuz it's not sending any movement packets if you not moving so it make things even harder to check.

Even worse, on 1.17-1.18 you can prob abuse duplicate packets to move and reach at the same times. So I don't think there any possible fix yet.

And for grim, they don't really want to add something that is going to false. I also really want to help but I don't think I'm good enough to fix something like this.

And yes moving back and fourth could possibly causing expand hitboxes but that something that has to be cuz If not it can cause falses.

**Comment by 0-x-2-2** [comment]
this only works while standing still and this is just from the uncertainty threshold

**Comment by ghost** [comment]
> this only works while standing still and this is just from the uncertainty threshold

Yes and that only works when standing still on 1.9 and will never happened on 1.8 without modifications.

**Comment by z7087** [comment]
Should there be a distinction between the collision box used for move checks and the hitbox used for the reach check?
If a client attacks a target, then it sends some transaction packets to update the target's location, and then attacks again, should the second attack not be checked?

**Comment by Hwaanx** [comment]
https://www.youtube.com/watch?v=bG-9m4Pvj1w

This is the second video

**Comment by Hwaanx** [comment]
also the reach check for newer versions false to a degree that some people get banned 1-3 falses atleast.

**Comment by 0-x-2-2** [comment]
> Yes and that only works when standing still on 1.9 and will never happened on 1.8 without modifications.

Not true, this happens silently all the time, and this is due to interpolation uncertainty. It is possible to increase the computational load and check slightly more accurately.

**Comment by ghost** [comment]
> > Yes and that only works when standing still on 1.9 and will never happened on 1.8 without modifications.
> 
> Not true, this happens silently all the time, and this is due to interpolation uncertainty. It is possible to increase the computational load and check slightly more accurately.

I'm talking about some other things ¯\_(ツ)_/¯

**Comment by AoElite** [comment]
@Hwaanx try latest build

**Comment by Hwaanx** [comment]
> @Hwaanx try latest build

I will test it now but i can say that latest grim reach check falses more.

**Comment by Hwaanx** [comment]
Okay, I have tested multiple times it still doesn't flag. (latest github)

**Comment by z7087** [comment]
Grim seems to combine all of an entity's past positions (up to the last player move) for movement checks, but the reach check may not require most of them

**Comment by AoElite** [comment]
@Hwaanx remember you should be moving, when you stand still there's only 1 idle packet being sent by the client per second instead of every tick

**Comment by Hwaanx** [comment]
ok so define said this is hard to fix but what can you say about false flags

**Comment by SamB440** [comment]
> ok so define said this is hard to fix but what can you say about false flags

I think part of the entity movement tracking relies on the prediction engine to determine if the player is reliably ticking in 1.9+. If simulation falses, then it might not detect unreliable ticking, which would false the reach check. 

Some uncertainty (e.g "Handle uncertainty of second transaction spanning over multiple ticks") could probably be removed in 1.19.4+ using bundle packets, as you are guaranteed the packets within a bundle will be processed on the same tick, but Grim supports all the way back to 1.8.

edit:
nvm the edit
